### ISSUE TITLE
ExtUtils::ParseXS needless deprecation of INCLUDE

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#10214]
Migrated from [rt.perl.org#73362](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=73362) (status was 'resolved')

Searchable as RT73362$

**Comment by p5pRT** [comment]
### From perlbug@plan9.de

#### Created by perlbug@plan9.de




For some reason\, somebody changed xsubpp to now deprecate functionality
that worked for decades \(well\, not quite\, but still a long time\)&#8203;:

&nbsp;   The INCLUDE directive with a command is deprecated\. Use INCLUDE\_COMMAND instead\! in cfperl\.xs\, line 2100

My question is\, \_is this REALLY necessary\_? Why cause maintainance
overhead for everybody using this feature\, when the existing feature is
working quite fine?

Please reconsider breaking things incomaptibly in needless cases like
these \(maybe it's not needless\, but I cannot at the moment imagine
anything breaking just because INCLUDE works like perl's builtin open
function\.\.\.\)

Thanks for reconsidering\, regardless of the outcome :\)

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=wishlist

Site configuration information for perl 5.10.0:

Configured by Marc Lehmann at Sat Feb 21 02:30:27 CET 2009.

Summary of my perl5 (revision 5 version 10 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.6.24-etchnhalf.1-amd64, archname=amd64-linux
    uname='linux cerebro 2.6.24-etchnhalf.1-amd64 #1 smp mon jul 21 10:36:02 utc 2008 x86_64 gnulinux '
    config_args='-Duselargefiles -Dxxxxuse64bitint -Uuse64bitall -Dusemymalloc=n -Dcc=gcc -Dccflags=-ggdb -gdwarf-2 -g3 -Dcppflags=-DPERL_ARENA_SIZE=16368 -D_GNU_SOURCE -I/opt/include -Doptimize=-O6 -msse2 -funroll-loops -fno-strict-aliasing -Dcccdlflags=-fPIC -Dldflags=-L/opt/perl/lib -L/opt/lib -Dlibs=-ldl -lm -lcrypt -Darchname=amd64-linux -Dprefix=/opt/perl -Dprivlib=/opt/perl/lib/perl5 -Darchlib=/opt/perl/lib/perl5 -Dvendorprefix=/opt/perl -Dvendorlib=/opt/perl/lib/perl5 -Dvendorarch=/opt/perl/lib/perl5 -Dsiteprefix=/opt/perl -Dsitelib=/opt/perl/lib/perl5 -Dsitearch=/opt/perl/lib/perl5 -Dsitebin=/opt/perl/bin -Dman1dir=/opt/perl/man/man1 -Dman3dir=/opt/perl/man/man3 -Dsiteman1dir=/opt/perl/man/man1 -Dsiteman3dir=/opt/perl/man/man3 -Dman1ext=1 -Dman3ext=3 -Dpager=/usr/bin/less -Uafs -Uusesfio -Uusenm -Uuseshrplib -Dd_dosuid -Dusethreads=undef -Duse5005threads=undef -Duseithreads=undef -Dusemultiplicity=undef -Demail=perl-binary@plan9.de -Dcf_email=perl-binary@plan9.de -Dcf_by=Marc Lehmann -Dlocincpth=/opt/perl/include /opt/include -Dmyhostname=localhost -Dmultiarch=undef -Dbin=/opt/perl/bin -Dxxxusedevel -DxxxDEBUGGING -Dxxxuse_debugging_perl -Dxxxuse_debugmalloc -des'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-ggdb -gdwarf-2 -g3 -fno-strict-aliasing -pipe -I/opt/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O6 -msse2 -funroll-loops -fno-strict-aliasing',
    cppflags='-DPERL_ARENA_SIZE=16368 -D_GNU_SOURCE -I/opt/include -ggdb -gdwarf-2 -g3 -fno-strict-aliasing -pipe -I/opt/include'
    ccversion='', gccversion='4.3.2', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags ='-L/opt/perl/lib -L/opt/lib -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64
    libs=-ldl -lm -lcrypt
    perllibs=-ldl -lm -lcrypt
    libc=/lib/libc-2.7.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.7'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O6 -msse2 -funroll-loops -fno-strict-aliasing -L/opt/perl/lib -L/opt/lib -L/usr/local/lib'

Locally applied patches:
    http://public.activestate.com/cgi-bin/perlbrowse/p/34209
    http://public.activestate.com/cgi-bin/perlbrowse/p/34507
    http://www.gossamer-threads.com/lists/perl/porters/232549
    embed.fnc:Perl_vcroak NULLOK


@INC for perl 5.10.0:
    /root/src/sex
    /opt/perl/lib/perl5
    /opt/perl/lib/perl5
    /opt/perl/lib/perl5
    /opt/perl/lib/perl5
    /opt/perl/lib/perl5
    .


Environment for perl 5.10.0:
    HOME=/root
    LANG (unset)
    LANGUAGE (unset)
    LC_CTYPE=en_US.UTF-8
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/root/s2:/root/s:/opt/bin:/opt/sbin:/bin:/sbin:/usr/bin:/usr/sbin:/usr/X11/bin:/usr/games:/usr/local/bin:/usr/local/sbin:/root/pserv:.
    PERL5LIB=/root/src/sex
    PERL5_CPANPLUS_CONFIG=/root/.cpanplus/config
    PERLDB_OPTS=ornaments=0
    PERL_ANYEVENT_DBI_TESTS=1
    PERL_ANYEVENT_EDNS0=1
    PERL_ANYEVENT_NET_TESTS=1
    PERL_ANYEVENT_PROTOCOLS=ipv4,ipv6
    PERL_ANYEVENT_STRICT=1
    PERL_BADLANG (unset)
    PERL_UNICODE=E
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @tsee

Hi Marc\,

perlbug @&#8203; plan9 \. de wrote&#8203;:
> \# New Ticket Created by  perlbug@&#8203;plan9\.de \# Please include the
> string&#8203;:  \[perl \#73362\] \# in the subject line of all future
> correspondence about this issue\. \# \<URL&#8203;:
> http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=73362 >

> For some reason\, somebody changed xsubpp to now deprecate
> functionality that worked for decades \(well\, not quite\, but still a
> long time\)&#8203;:
> 
> The INCLUDE directive with a command is deprecated\. Use
> INCLUDE\_COMMAND instead\! in cfperl\.xs\, line 2100
> 
> My question is\, \_is this REALLY necessary\_? Why cause maintainance 
> overhead for everybody using this feature\, when the existing feature
> is working quite fine?

That would be me\. A quick note before I get into the details&#8203;: I do not
intend to \*break\* things any time soon but rather warn about potential
problems\.

The problem with the existing feature is that it's very easy to break
accidentally\. For example\, a common use case is to call into the perl
that's being used to build the given module\. To the best of my
knowledge\, that wasn't possible before\. You'd just run COMMAND&#8203;: perl \.\.\.
| which could be some random perl that's in your path\. This broke builds
of Wx\, for example\. INCLUDE\_COMMAND is slightly magical in this regard\.
You can use $^X to refer to the currently running \(i\.e\. xsubpp\-running\)
perl\.

> Please reconsider breaking things incomaptibly in needless cases like
>  these \(maybe it's not needless\, but I cannot at the moment imagine 
> anything breaking just because INCLUDE works like perl's builtin open
>  function\.\.\.\)
> 
> Thanks for reconsidering\, regardless of the outcome :\)

Okay\, you'll ask rightly why I didn't just modify INCLUDE&#8203;: \.\.\. |
instead\. The answer is two\-fold\.
a\) I think Wx is not the only module that's affected by this lurking bug 
\(INCLUDE&#8203;: perl \-e \.\.\. |\) and I consider warning about it a good idea\. 
Maybe a more specific warning for /INCLUDE&#8203;:\\s\*perl/ would have been an 
even better idea\.
b\) A posse of porters on \#p5p ganged up on me and strongly voted in 
favour of separating the two uses of INCLUDE\. I let them convince me 
that it's cleaner\.

Best regards\,
Steffen

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, Mar 08\, 2010 at 09&#8203;:11&#8203;:20PM \+0100\, Steffen Mueller wrote&#8203;:
> The problem with the existing feature is that it's very easy to break
> accidentally\. For example\, a common use case is to call into the perl
> that's being used to build the given module\. To the best of my
> knowledge\, that wasn't possible before\. You'd just run COMMAND&#8203;: perl \.\.\.
> | which could be some random perl that's in your path\. This broke builds
> of Wx\, for example\. INCLUDE\_COMMAND is slightly magical in this regard\.
> You can use $^X to refer to the currently running \(i\.e\. xsubpp\-running\)
> perl\.

Hmm\,\. looks like perlxstut\.pod nneds updating to mention INCLUDE\_COMMAND
and deprecate COMMAND \.\.\. |

\-\- 
Please note that ash\-trays are provided for the use of smokers\,
whereas the floor is provided for the use of all patrons\.
&nbsp;    \-\- Bill Royston

**Comment by p5pRT** [comment]
### From @tsee

Dave Mitchell wrote&#8203;:
> On Mon\, Mar 08\, 2010 at 09&#8203;:11&#8203;:20PM \+0100\, Steffen Mueller wrote&#8203;:
>> The problem with the existing feature is that it's very easy to break
>> accidentally\. For example\, a common use case is to call into the perl
>> that's being used to build the given module\. To the best of my
>> knowledge\, that wasn't possible before\. You'd just run COMMAND&#8203;: perl \.\.\.
>> | which could be some random perl that's in your path\. This broke builds
>> of Wx\, for example\. INCLUDE\_COMMAND is slightly magical in this regard\.
>> You can use $^X to refer to the currently running \(i\.e\. xsubpp\-running\)
>> perl\.
> 
> Hmm\,\. looks like perlxstut\.pod nneds updating to mention INCLUDE\_COMMAND
> and deprecate COMMAND \.\.\. |

I think the doc patching was postponed due to the 5\.12 freeze\. The 
version of EU&#8203;::ParseXS that does the INCLUDE\_COMMAND isn't part of 5\.12 
as far as I know and the EU&#8203;::PXS docs are effectively in perlxs\.pod \(and 
to some extent in perlxstut\)\.

Marc has a valid point\. If the executed program is not perl\, 
INCLUDE\_COMMAND has not benefit over INCLUDE&#8203;: \.\.\. | except it's a little 
clearer\.

Marc\, if we changed the warning to just appear in case of "INCLUDE&#8203;: perl 
\.\.\. |" and to more specifically point to the potentially false 
assumption about "perl" being "$^X"\, would that be an acceptable 
solution for you? Of course\, this would mean that INCLUDE&#8203;: \.\.\. | would 
not be deprecated but just slightly discouraged \(via documentation\) in 
favour of INCLUDE\_COMMAND\.

Nicholas\, back in the day\, you uttered mild disgust at the "INCLUDE&#8203;: \.\.\. 
|" construct\. Would this trade\-off be fine with you\, too?

Cheers\,
Steffen

**Comment by p5pRT** [comment]
### From @nwc10

On Wed\, Mar 10\, 2010 at 02&#8203;:09&#8203;:41PM \+0100\, Steffen Mueller wrote&#8203;:

> Marc\, if we changed the warning to just appear in case of "INCLUDE&#8203;: perl 
> \.\.\. |" and to more specifically point to the potentially false 
> assumption about "perl" being "$^X"\, would that be an acceptable 
> solution for you? Of course\, this would mean that INCLUDE&#8203;: \.\.\. | would 
> not be deprecated but just slightly discouraged \(via documentation\) in 
> favour of INCLUDE\_COMMAND\.
> 
> Nicholas\, back in the day\, you uttered mild disgust at the "INCLUDE&#8203;: \.\.\. 
> |" construct\. Would this trade\-off be fine with you\, too?

Yes\. \(I can't remember particularly why \- I think it's because I thought
"It's a file"\. It's somewhat the same reaction I have to the idea that PDFs
would want to contain JavaScript\)

Seems a very good compromise\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @tsee

Hi Marc\, hi all\,

Nicholas Clark wrote&#8203;:
> On Wed\, Mar 10\, 2010 at 02&#8203;:09&#8203;:41PM \+0100\, Steffen Mueller wrote&#8203;:
> 
>> Marc\, if we changed the warning to just appear in case of "INCLUDE&#8203;: perl 
>> \.\.\. |" and to more specifically point to the potentially false 
>> assumption about "perl" being "$^X"\, would that be an acceptable 
>> solution for you? Of course\, this would mean that INCLUDE&#8203;: \.\.\. | would 
>> not be deprecated but just slightly discouraged \(via documentation\) in 
>> favour of INCLUDE\_COMMAND\.
>>
>> Nicholas\, back in the day\, you uttered mild disgust at the "INCLUDE&#8203;: \.\.\. 
>> |" construct\. Would this trade\-off be fine with you\, too?
> 
> Yes\. \(I can't remember particularly why \- I think it's because I thought
> "It's a file"\. It's somewhat the same reaction I have to the idea that PDFs
> would want to contain JavaScript\)
> 
> Seems a very good compromise\.

This compromise will be in an upcoming ExtUtils&#8203;::ParseXS release thanks 
to David Golden's quick reaction\.

Best regards\,
Steffen

**Comment by p5pRT** [comment]
### From @tsee

Marking this ticket as resolved as the described workaround has been
implemented in ExtUtils&#8203;::ParseXS 0\.2205 \(on CPAN\)\. \-\-Steffen

**Comment by p5pRT** [comment]
@tsee - Status changed from 'open' to 'resolved'
